syntax = "proto3";

service GameCore {
    rpc ConnectClient(TPingRequest) returns (TPingResponse) {}

    rpc WaitInQueue(TPingRequest) returns (TPingResponse) {}

    rpc DisconnectClient(TPingRequest) returns (TPingResponse) {}

    rpc SubscribeForNotifications(TPingRequest) returns (stream TSystemNotification) {}

    rpc SessionMove(TSessionMoveRequest) returns (TSessionMoveResponse) {}
}

message TPingRequest {
    string username = 1;
}

message TPingResponse {
    optional string message = 1;
}

message TSystemNotification {
    enum NotificationType {
        REGULAR_MESSAGE = 0;
        SESSION_INFO_MESSAGE = 1;
        TURN_INFO_MESSAGE = 2;
        RESULT_MESSAGE = 3;
    }
    enum Role {
        MAFIA_ROLE = 0;
        CITIZEN_ROLE = 1;
        COMMISSAR_ROLE = 2;
    }
    message SessionInfo {
        string session_id = 1;
        Role role = 2; 
        repeated string players = 3;
        string all_key = 4;
        optional string mafia_key = 5;
    }
    message TurnInfo {
        Role turn = 1;
        repeated string vote_options = 2;
        optional string target_username = 3;
        optional Role target_role = 4;
    }
    message SessionResult {
        message Client {
            string username = 1;
            bool alive = 2;
            optional Role role = 3;
        }
        bool citizens_wins = 1;
        repeated Client clients = 2;
    }

    NotificationType type = 1;
    oneof data {
        string message = 2;
        SessionInfo session_info = 3;  
        TurnInfo turn_info = 4;
        SessionResult result = 5;
    }
}

message TSessionMoveRequest {
    string username = 1;
    string vote_for = 2;
    string session_id = 3;
}

message TSessionMoveResponse {
    optional string message = 1;
}
